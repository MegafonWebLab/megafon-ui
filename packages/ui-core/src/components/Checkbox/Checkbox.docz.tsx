import * as React from 'react';

interface ICheckboxWrapperProps {
    children: (prop: { checked: boolean; onChange: () => void }) => JSX.Element;
}

export const LightColorWrapper: React.FC = ({ children }) => {
    return (
        <div style={{ backgroundColor: '#00b956' }}>
            {children}
        </div>
    );
};

export const CheckboxWrapper: React.FC<ICheckboxWrapperProps> = (props) => {
    const [checked, setChecked] = React.useState<boolean>(false);

    const handleChange = () => {
        setChecked(!checked);
    };

    return (
        <div style={{ padding: '20px 10px 8px 10px' }}>
            {props.children({
                checked,
                onChange: handleChange,
            })}
        </div>
    );
};
