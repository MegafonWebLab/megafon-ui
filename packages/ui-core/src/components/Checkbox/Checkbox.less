@import (reference) '~styles/base.less';

@b: e('.mfui-checkbox');

@{b} {
    margin: 0 5px 12px 5px;

    &__inner {
        position: relative;

        display: inline-block;
    }

    &__label {
        display: inline;

        cursor: pointer;
    }

    &__input {
        position: absolute;
        top: 0;
        left: 0;
        z-index: -1;

        width: 0;
        height: 0;

        opacity: 0;
    }

    &__custom-input {
        position: absolute;
        left: 0;

        display: flex;
        align-items: center;
        justify-content: center;
        box-sizing: border-box;
        width: 20px;
        height: 20px;
        border: 2px solid;
        border-radius: 2px;
    }

    &__icon {
        display: block;
        width: 32px;
        height: 32px;

        opacity: 0;

        user-select: none;
    }

    &__extra-content {
        position: relative;
        left: 8px;

        display: inline-block;
        margin-top: -2px;
        margin-right: -20px;

        vertical-align: middle;
    }

    &_checked {
        @{b} {
            &__custom-input {
                border: none;
            }

            &__icon {
                opacity: 1;
            }
        }
    }

    &_font-size_regular {
        font-size: 15px;
        line-height: 24px;

        @{b} {
            &__inner {
                padding: 4px 15px 4px 35px;
            }

            &__custom-input {
                top: 6px;
            }
        }
    }

    &_font-size_small {
        font-size: 12px;
        line-height: 16px;

        @{b} {
            &__inner {
                padding: 1px 15px 4px 35px;
            }

            &__custom-input {
                top: 0;
            }
        }
    }

    &_color_dark {
        @{b} {
            &__custom-input {
                border-color: var(--spbSky2);
            }

            &__label_no-touch:hover @{b}__custom-input {
                border-color: var(--buttonHoverGreen);
            }

            &__label:active @{b}__custom-input {
                border-color: var(--buttonDown) !important;

                background-color: var(--buttonDown);
            }

            &__icon {
                fill: var(--stcWhite);
            }
        }
    }

    &_color_dark&_checked {
        @{b} {
            &__custom-input {
                border-color: var(--brandGreen);

                color: var(--stcWhite);

                background-color: var(--brandGreen);
            }

            &__label_no-touch:hover @{b}__custom-input {
                background-color: var(--buttonHoverGreen);
            }

            &__label:active @{b}__custom-input {
                background-color: var(--buttonDown) !important;
            }
        }
    }

    &_color_dark&_checked&_disabled {
        @{b} {
            &__custom-input,
            &__label:active @{b}__custom-input {
                background-color: var(--spbSky2);
            }
        }
    }

    &_color_dark&_disabled {
        @{b} {
            &__label {
                cursor: default;

                pointer-events: none;
            }

            &__custom-input,
            &__label:active @{b}__custom-input {
                border-color: var(--spbSky2) !important;

                background-color: var(--spbSky1);
            }
        }
    }

    &_color_dark&_error {
        @{b} {
            &__custom-input {
                border-color: var(--fury);
            }
        }
    }

    &_color_light {
        color: var(--stcWhite);

        @{b} {
            &__custom-input {
                border-color: var(--stcWhite);

                color: var(--brandGreen);

                background-color: var(--stcWhite);
            }
        }
    }
}
