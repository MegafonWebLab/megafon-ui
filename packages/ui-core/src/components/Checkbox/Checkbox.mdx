---
name: Checkbox
menu: Компоненты
route: components/checkbox
---

import { Playground, Props } from 'docz';
import Checkbox from './Checkbox';
import Info from 'icons/System/16/Info_16.svg';
import { CheckboxWrapper, LightColorWrapper } from './Checkbox.docz';

# Checkbox

## Properties

<Props of={Checkbox} />

## Basic usage

<Playground>
    <CheckboxWrapper>
        {({ checked, onChange }) =>
            <Checkbox checked={checked} onChange={onChange}>
                I agree
            </Checkbox>
        }
    </CheckboxWrapper>
</Playground>

## Checkbox with small label size

<Playground>
    <CheckboxWrapper>
        {({ checked, onChange }) =>
            <Checkbox
                checked={checked}
                onChange={onChange}
                fontSize="small"
            >
                I agree
            </Checkbox>
        }
    </CheckboxWrapper>
</Playground>

## Checkbox usage with error

<Playground>
    <CheckboxWrapper>
        {({ checked, onChange }) =>
            <Checkbox checked={checked} onChange={onChange} error>
                I agree
            </Checkbox>
        }
    </CheckboxWrapper>
</Playground>

## Checkbox usage for light color scheme

<Playground>
    <LightColorWrapper>
        <CheckboxWrapper>
            {({ checked, onChange }) =>
                <Checkbox
                    checked={checked}
                    onChange={onChange}
                    color="light"
                >
                    I agree
                </Checkbox>
            }
        </CheckboxWrapper>
    </LightColorWrapper>
</Playground>

## Checkbox with extraContent

<Playground>
    <CheckboxWrapper>
        {({ checked, onChange }) =>
            <Checkbox
                checked={checked}
                onChange={onChange}
                extraContent={
                    <div style={{ backgroundColor: '#00B956', color: '#FFF', padding: '5px'}}>
                        Some Extra Content
                    </div>
                }
            >
                I agree
            </Checkbox>
        }
    </CheckboxWrapper>
</Playground>

## Disabled unchecked checkbox

<Playground>
    <CheckboxWrapper>
        {() => <Checkbox disabled>I agree</Checkbox>}
    </CheckboxWrapper>
</Playground>

## Disabled checked checkbox

<Playground>
    <CheckboxWrapper>
        {() => <Checkbox checked disabled>I agree</Checkbox>}
    </CheckboxWrapper>
</Playground>
