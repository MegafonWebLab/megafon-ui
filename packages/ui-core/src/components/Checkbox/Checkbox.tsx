import * as PropTypes from 'prop-types';
import * as React from 'react';
import cnCreate from 'utils/cnCreate';
import './Checkbox.less';
import CheckedIcon from 'icons/System/16/Checked_16.svg';
import detectTouch from 'utils/detectTouch';
import filterDataAttrs, { IDataAttributes } from '../../utils/dataAttrs';

export interface ICheckboxProps extends IDataAttributes {
    /** Цвет чекбокса */
    color?: 'dark' | 'light';
    /** Дополнительный класс корневого элемента */
    className?: string;
    /** Размер текста */
    fontSize?: 'regular' | 'small';
    /** Имя компонента в DOM */
    name?: string;
    /** Значение чекбокса */
    value?: string;
    /** Управление состоянием вкл/выкл компонента */
    checked?: boolean;
    /** Управление возможностью взаимодействия с компонентом */
    disabled?: boolean;
    /** Отобразить компонент в состоянии ошибки */
    error?: boolean;
    /** Дополнительный контент справа */
    extraContent?: JSX.Element[] | Element[] | JSX.Element | Element | string;
    children: JSX.Element[] | Element[] | JSX.Element | Element | string;
    /** Обработчик изменения значения */
    onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const cn = cnCreate('mfui-beta-checkbox');
class Checkbox extends React.Component<ICheckboxProps, {}> {
    static propTypes = {
        className: PropTypes.string,
        fontSize: PropTypes.oneOf(['regular', 'small']),
        color: PropTypes.oneOf(['dark', 'light']),
        name: PropTypes.string,
        value: PropTypes.string,
        checked: PropTypes.bool,
        disabled: PropTypes.bool,
        error: PropTypes.bool,
        children: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
            PropTypes.string,
        ]),
        extraContent: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
            PropTypes.string,
        ]),
        dataAttrs: PropTypes.objectOf(PropTypes.string.isRequired),
        onChange: PropTypes.func,
    };

    static defaultProps: Partial<ICheckboxProps> = {
        color: 'dark',
        fontSize: 'regular',
        checked: false,
    };

    isTouch: boolean = detectTouch();

    handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
        const { onChange } = this.props;
        onChange && onChange(e);
    }

    render() {
        const {
            className,
            fontSize,
            name,
            color,
            value,
            checked,
            disabled,
            error,
            children,
            extraContent,
            dataAttrs,
        } = this.props;

        return (
            <div
                className={cn(
                    '',
                    {
                        'font-size': fontSize,
                        color,
                        checked,
                        disabled,
                        error,
                    },
                    className
                )}
                {...filterDataAttrs(dataAttrs)}
            >
                <div className={cn('inner')}>
                    <label className={cn('label', { 'no-touch': !this.isTouch })}>
                        <input
                            className={cn('input')}
                            type="checkbox"
                            name={name}
                            value={value}
                            checked={checked}
                            onChange={this.handleChange}
                            disabled={disabled}
                        />
                        <div className={cn('custom-input')}>
                            <CheckedIcon className={cn('icon')} />
                        </div>
                        {children}
                    </label>
                    {extraContent && <span className={cn('extra-content')}>{extraContent}</span>}
                </div>
            </div>
        );
    }
}

export default Checkbox;
