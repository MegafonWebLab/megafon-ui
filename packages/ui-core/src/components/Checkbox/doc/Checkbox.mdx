---
name: Checkbox
menu: Компоненты
route: components/checkbox
---

import { Playground, Props } from 'docz';
import Checkbox from '../Checkbox';
import Example from './Checkbox.example.mdx';

# Checkbox

<CopyToClipboardBox text="import { Checkbox } from '@megafon/ui-core';" />

<DoczTabs>
    <Example />
    <Props of={Checkbox} />
</DoczTabs>
