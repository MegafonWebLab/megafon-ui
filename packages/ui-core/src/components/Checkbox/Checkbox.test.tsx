import * as React from 'react';
import { shallow } from 'enzyme';
import Checkbox, { ICheckboxProps } from './Checkbox';

const props = {
    className: 'testClassName',
    name: 'testInputName',
    value: 'testInputValue',
    checked: true,
    disabled: true,
    error: true,
    extraContent: 'text',
};

const dataAttrs: ICheckboxProps['dataAttrs'] = {
    root: {
        'data-test': 'test',
        'incorrect-attr': 'test',
    },
    input: {
        'data-test': 'test',
        'incorrect-attr': 'test',
    },
    customInput: {
        'data-test': 'test',
        'incorrect-attr': 'test',
    },
    extraContent: {
        'data-test': 'test',
        'incorrect-attr': 'test',
    },
};

describe('<Checkbox />', () => {
    describe('layout', () => {
        it('renders Checkbox', () => {
            const wrapper = shallow(<Checkbox>Тестовая строка</Checkbox>);
            expect(wrapper).toMatchSnapshot();
        });

        it('renders Checkbox with props', () => {
            const wrapper = shallow(
                <Checkbox {...props} fontSize="small" color="light" onChange={jest.fn()}>
                    Тестовая строка
                </Checkbox>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        it('renders Checkbox on mobile resolution', () => {
            const onTouchStart = window.ontouchstart;
            window.ontouchstart = jest.fn();

            const wrapper = shallow(<Checkbox>Тестовая строка на мобильном разрешении</Checkbox>);
            expect(wrapper).toMatchSnapshot();

            window.ontouchstart = onTouchStart;
        });

        it('renders with data attrs', () => {
            const wrapper = shallow(<Checkbox dataAttrs={dataAttrs}>Тестовая строка</Checkbox>);
            expect(wrapper).toMatchSnapshot();
        });

        it('renders with classes', () => {
            const wrapper = shallow(
                <Checkbox classes={{ icon: 'custom-icon', inner: 'custom-inner' }}>Тестовая строка</Checkbox>,
            );
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('handleChange', () => {
        it('calls onChange on change input', () => {
            const event = {
                type: 'change',
                nativeEvent: {},
            } as React.ChangeEvent;
            const checked = false;
            const handleChange = jest.fn();

            const wrapper = shallow(
                <Checkbox onChange={handleChange} checked={checked}>
                    Тестовая строка
                </Checkbox>,
            );

            wrapper.find('input').simulate('change', event);
            expect(handleChange).toBeCalledWith(!checked);
        });

        it('calls onChange on Enter click', () => {
            const event = {
                type: 'keydown',
                nativeEvent: {
                    code: 'Enter',
                },
            } as React.KeyboardEvent;
            const checked = false;
            const handleChange = jest.fn();

            const wrapper = shallow(
                <Checkbox onChange={handleChange} checked={checked}>
                    Тестовая строка
                </Checkbox>,
            );

            wrapper.find('div[role="checkbox"]').simulate('keydown', event);
            expect(handleChange).toBeCalledWith(!checked);
        });

        it('not calls onChange on keydown, if keyboard code is not Enter', () => {
            const event = {
                type: 'keydown',
                nativeEvent: {
                    code: 'Tab',
                },
            } as React.KeyboardEvent;
            const checked = false;
            const handleChange = jest.fn();

            const wrapper = shallow(
                <Checkbox onChange={handleChange} checked={checked}>
                    Тестовая строка
                </Checkbox>,
            );

            wrapper.find('div[role="checkbox"]').simulate('keydown', event);
            expect(handleChange).not.toBeCalled();
        });
    });
});
