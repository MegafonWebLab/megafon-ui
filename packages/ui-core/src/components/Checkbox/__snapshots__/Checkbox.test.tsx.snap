// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render Checkbox 1`] = `
<div>
  <div
    class="mfui-checkbox mfui-checkbox_font-size_regular mfui-checkbox_color_dark"
  >
    <div
      class="mfui-checkbox__inner"
    >
      <label
        class="mfui-checkbox__label mfui-checkbox__label_no-touch"
      >
        <input
          class="mfui-checkbox__input"
          tabindex="-1"
          type="checkbox"
          value=""
        />
        <div
          aria-checked="false"
          class="mfui-checkbox__custom-input"
          role="checkbox"
          tabindex="0"
        >
          <test-file-stub
            classname="mfui-checkbox__icon"
          />
        </div>
        Тестовая строка
      </label>
      <span
        class="mfui-checkbox__extra-content"
      >
        Extra content
      </span>
    </div>
  </div>
</div>
`;
